libvast/builtins/formats/json.cpp

```diff
@@ -0,0 +1,356 @@
+
+#include <vast/adaptive_table_slice_builder.hpp>
+#include <vast/arrow_table_slice.hpp>
+#include <vast/concept/parseable/numeric/bool.hpp>
+#include <vast/concept/parseable/vast/data.hpp>
+#include <vast/concept/printable/vast/json.hpp>
+#include <vast/defaults.hpp>
+#include <vast/detail/assert.hpp>
+#include <vast/generator.hpp>
+#include <vast/operator_control_plane.hpp>
+#include <vast/plugin.hpp>
+
+#include <arrow/record_batch.h>
+
+#include <simdjson.h>
+
+namespace vast::plugins::json {
+
+namespace {
+
+using parse_result = std::optional<simdjson::error_code>;
+
+class parser_buffer {
+public:
+  void append(std::string_view json) {
+    if (const auto available_bytes = capacity_ - end_;
+        available_bytes < json.size()) {
+      const auto bytes_missing = json.size() - available_bytes;
+      if (begin_ >= bytes_missing) {
+        std::shift_left(buffer_.get() + begin_, buffer_.get() + end_, begin_);
+        end_ -= begin_;
+        begin_ = 0u;
+      } else {
+        auto new_buffer = std::make_unique<char[]>(
+          end_ + simdjson::SIMDJSON_PADDING + json.size());
+        std::copy(buffer_.get(), buffer_.get() + end_, new_buffer.get());
+        // The simdjson suggests to initialize the padding part to either 0s or
+        // spaces.
+        std::fill_n(new_buffer.get() + capacity_, simdjson::SIMDJSON_PADDING,
+                    '\0');
+        capacity_ = end_ + json.size();
+        buffer_ = std::move(new_buffer);
+      }
+    }
+    std::copy(json.begin(), json.end(), buffer_.get() + end_);
+    end_ += json.size();
+  }
+
+  std::string_view view() {
+    return {buffer_.get() + begin_, end_ - begin_};
+  }
+
+  operator bool() {
+    return end_ != 0u;
+  }
+
+  // Reuse the already allocated buffer.
+  void reset() {
+    end_ = 0u;
+    begin_ = 0u;
+  }
+  // The input json can contain the whole event and a part of the other one. E.g
+  // the input chunk is: {"a":5}{"a" This whole string will be in the buffer_
+  // and the parser will tell us that it only consumed the first event. The
+  // {"a":5} was consumed and we want to add the next chunk to the leftover
+  // part. We set begin_ at the end of the consumed events because our allocated
+  // buffer might still have enough capacity to write leftover + new chunk.
+  void truncate(std::size_t n) {
+    begin_ = end_ - n;
+  }
+
+private:
+  std::unique_ptr<char[]> buffer_;
+  std::size_t begin_{0u};
+  std::size_t end_{0u};
+  // The available bytes size for json string. Doesn't include the required
+  // simdjson padding.
+  std::size_t capacity_{0u};
+};
+
+parse_result
+parse_impl(simdjson::ondemand::value val, auto& pusher, size_t depth);
+
+parse_result parse_number(simdjson::ondemand::value val, auto& pusher) {
+  switch (val.get_number_type().value_unsafe()) {
+    case simdjson::ondemand::number_type::floating_point_number: {
+      auto result = val.get_double();
+      if (result.error())
+        return result.error();
```

Comment:
Not sure. Look at the weird comment in this function.
https://github.com/simdjson/simdjson/blob/master/singleheader/simdjson.h#L11866

I would say given "2153\n" would yield double_type but would fail on the get_double()

I have tested

```
TEST(WTF)
{
  constexpr auto json = std::string_view{R"({"ble": 46363643;})"};
  simdjson::ondemand::parser p;
  simdjson::padded_string k{json};
  for (auto s : p.iterate_many(k).value())
  {
    auto obj = s.get_object().value_unsafe();
    for (auto p : obj)
    {
      if (p.error())
      {
        MESSAGE("p err");
      }
      MESSAGE(p.unescaped_key().value_unsafe());
      auto v = p.value();
      auto type = v.get_number_type().value_unsafe();
      MESSAGE((int)type);
      auto dbl = v.get_double();
      if (dbl.error())
      {
        MESSAGE("dbl error");
      }
      else
      {
      MESSAGE(dbl.value_unsafe());

      }
    }
  }
}
```

And the output is 
```
 -> [98, 108, 101] [line 345]
  -> 1 [line 348]
  -> dbl error [line 352]
```
Proving that it failed :/

Reply:
Okay, thanks for confirming. Then let's keep this as-is. :|